spec/controllers/training_modules_users_controller_spec.rb

```diff
@@ -30,6 +33,17 @@
           .to be_between(1.minute.ago, 1.minute.from_now)
       end
     end
-  end
+    context 'nonexistent slide' do
+      let(:nonexistent_slide_params) do
+        { module_id: t_mod.slug,
+          slide_id: 'not-a-real-slide',
+          user_id: user.id }
+      end
 
+      it 'renders a response' do
+        post :create_or_update, nonexistent_slide_params, format: :json
+        expect(response).to be
```

Comment:
I think the test makes sense for the current controller behavior. The question is whether the controller behavior is right.

I think it is, because this controller is basically the endpoint for an ajax request that says 'I made it to this slide', and the response it expects back is 'Okay, you made it to that slide'. Current implementation is to reply 'Okay, you made it to null-slide.' It's up to the client to figure out what to do from there. There's already another route by which the client is similultaneously being redirected to the 404 page, but this controller doesn't know or care.

It also makes for a pretty clean method: the render happens the same way for existing and nonexisting slides, and then it just returns before the business logic if the guard clause gets triggered.